generate/avro.go

```
@@ -0,0 +1,344 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse generate go file %q: %v", path, err)
+	}
+	if !ast.FilterFile(goFile, func(itemName string) bool { return itemName == name }) {
+		return nil, fmt.Errorf("a struct named %q was not found in file %q", name, path)
+	}
+
+	if length := len(goFile.Decls); length != 1 {
+		return nil, fmt.Errorf("failed to filter declarations to a single one named %q, found %d", name, length)
+	}
+
+	d, ok := goFile.Decls[0].(*ast.GenDecl)
+	if !ok {
+		return nil, errors.New("filtered declaration is of unknown ast type")
+	}
+	if len(d.Specs) != 1 {
+		return nil, errors.New("unexpected number of specs in declaration")
+	}
+	t, ok := d.Specs[0].(*ast.TypeSpec)
+	if !ok {
+		return nil, errors.New("filtered type is of unknown ast type")
+	}
+
+	return t, nil
+}
+
+func parseStructTag(literal *ast.BasicLit) (name, description string, omitEmpty bool) {
+	if literal == nil {
+		return
+	}
+	tag := reflect.StructTag(strings.Trim(literal.Value, "`"))
+
+	description = tag.Get("description")
+	jsonValue := tag.Get("json")
+	jsonSplits := strings.Split(jsonValue, ",")
+	name = jsonSplits[0]
+	if len(jsonSplits) > 1 {
+		for _, split := range jsonSplits[1:] {
+			if strings.ToLower(split) == "omitempty" {
+				omitEmpty = true
+				return
+			}
+		}
+	}
+	return
+}
+
+// writeAvroStruct returns an apply function intended to be called for the start of each node.
+// The corresponding Post function will be called after the children of the node have been traversed.
+func writeAvroStruct(cfg avroConfig, name, defaultFields string) astutil.ApplyFunc {
+	return func(c *astutil.Cursor) bool {
+		if c.Name() == "Node" {
+			var namespace string
+			if len(cfg.namespace) != 0 {
+				namespace = fmt.Sprintf(`"namespace":%q,`, strings.Join(cfg.namespace, "."))
+			}
+			if _, err := fmt.Fprintf(cfg.writer, `"name":%q,%s"type":"record","fields":[%s`, name, namespace, defaultFields); err != nil {
```

Is there some way we can report this error?

---

Not without making it some sort of type and making this a method of that type which seems pretty complicated. What will end up happening right now is the JSON will be invalid and fail later parsing.